### ISSUE TITLE
FEEL expressions for Camunda attributes / entities.

### ISSUE THREAD

**Issue by rob2universe** [camunda/zeebe#10987]
**Is your feature request related to a problem? Please describe.**

- Feature parity, flexibility
- While a businessKey could be used (if one created a process data for it), it is still commonly used to processInstanceKey as a technical correlation criteria or reference, be it to match message or to include the information in outgoing (email) notifications.
-I can replace this with a process data, but then need to think about how to init the process data with unique Id.  Having a FEEL expression processInstanceKey in an out data mapping on the trigger makes that easy (the upcoming random number() would also be helpful, but carries no meaning).
- Especially if we start working with connector functions, we want to be able to map such data into a connector for...whatever the connector does. Here we don't have the programmatic access to the execution information anymore. If a connector creates a job on another system, then it is nice to carry the technical references to Camunda around
- Since Camunda currently does not support the processInstanceKey as a query parameter, a workaround is to copy it into a data and then query by this data. However, this workaround would work better if it could be done in an expression and would not require  a service task.

**Describe the solution you'd like**

offer FEEL expressions like: processInstanceKey, processName, parentProcessInstanceKey, flowNode,...
to be used in FEEL expressions such as data mappings 

**Describe alternatives you've considered**

Implement a service task, which reads information from the job and write it back to the process. However, this only works for the information available in the current scope, not for a future flow node scope (next service task).

**Comment by korthout** [comment]
Thanks for reporting @rob2universe 

@felix-mueller Does this feature make sense from the PM's perspective? If so, I could specify how we could support this in Zeebe.

**Comment by skayliu** [comment]
Look forward to this feature.

**Comment by skayliu** [comment]
Look forward to this feature.

**Comment by MaxTru** [comment]
+1 for this feature from the Connectors team. We also retrieved user requests: https://github.com/camunda/connector-sdk/issues/137

**Comment by felix-mueller** [comment]
@korthout feature generally makes sense from my perspective.

**Comment by korthout** [comment]
We already have [a technical concept](https://github.com/camunda/zeebe/blob/clients/go/v8.1.5/expression-language/src/main/java/io/camunda/zeebe/el/EvaluationContext.java#L25-L38) implemented to provide additional properties to the evaluation of FEEL expressions inside Zeebe. This is already used to provide the multi-instance properties when evaluating the [completion condition](https://docs.camunda.io/docs/components/modeler/bpmn/multi-instance/#completion-condition). Note that this hides/shadows process variables with these names from the expression. The concept is built to be reused in other parts of our code.

IMO, the biggest challenges are product decisions:
1. to carefully consider which properties should be available (e.g. `processInstanceKey` seems like a reasonable candidate, but I've also heard the counter-argument that a business key should be preferred over technical keys; in other words, we should not leak these technical concepts);
2. with what names these properties should be available (i.e. if we use `processIntanceKey` and the user already has a process variable with the same name, then this variable's value will not be available in the expression because it is shadowed by the system-provided property);
3. and in which [expressions](https://docs.camunda.io/docs/next/components/concepts/expressions/) these properties should be provided (e.g. input mappings, output mappings, sequence flow conditions, [called decision id](https://docs.camunda.io/docs/next/components/modeler/bpmn/business-rule-tasks/#defining-a-called-decision), or even inside the evaluation of decisions).

@felix-mueller We need to find answers to these questions before we can do a technical breakdown of the work. I suggest we consider this topic on the product roadmap, consider the priority, and look for answers to these questions at/after a kickoff.

**Comment by korthout** [comment]
I've worked out the ideas of this feature in more detail in https://github.com/camunda/zeebe/issues/11459, which is pretty much the same feature requested in this issue but then specifically for the caught Error in a BPMN Error Catch Event and making that available in the output mappings.

That issue can function as an excellent example of finding the answer to the design question:

  2. **with what names should these properties be available** (see my [previous comment](https://github.com/camunda/zeebe/issues/10987#issuecomment-1369730200)). We should carefully consider naming collisions to avoid breaking user space

Once that issue is completed, we can more easily work on this issue. @felix-mueller I think this issue should be considered for the product roadmap.

**Comment by korthout** [comment]
The team has decided on the namespace to use for expression contexts (see #11459): `camunda`

These attributes and entities could be made available in many ways within this namespace. For example, we could make the `processInstanceKey` available via:
- `camunda.processInstanceKey`
- `camunda.processInstance.key`
- `camunda.process.instance.key`
- `camunda.element.processInstanceKey`
- `camunda.flowNode.processInstanceKey`
- etc.

Note that other features have already reserved some expression contexts:
- `camunda.variables` -> all visible process variables
- `camunda.local` -> local variables only
- `camunda.multi` -> multi-instance properties
- `camunda.error` -> details about a caught BPMN error
- `camunda.env` -> environment variables

What ever we choose, it should play nicely with the other expression contexts

**Comment by yT0n1** [comment]
@korthout any plans to add this to the roadmap?

**Comment by korthout** [comment]
@yT0n1 Absolutely, thanks for the reminder though 🙇 I've reached out to PM to learn how to get this on the roadmap properly.

**Comment by korthout** [comment]
@yT0n1 PM has created https://github.com/camunda/product-hub/issues/1909 to track this internally. Thanks again for bringing it back to my attention.

**Comment by rb83** [comment]
This would be useful in our case as well. I'm not sure if the product-hub issue link given works for public users, but it's good to know this is in the works.

**Comment by gevezex** [comment]
> @yT0n1 PM has created https://github.com/camunda/product-hub/issues/1909 to track this internally. Thanks again for bringing it back to my attention.

This url returns a 404

**Comment by korthout** [comment]
>This url returns a 404

@gevezex camunda/product-hub is an internal repository. A corresponding issue is typically created in https://github.com/camunda/issues/issues but that hasn't happened for this issue yet.

**Comment by michaelarnauts** [comment]
I'm also looking forward to these variables.

I need to generate an UUID (to send in a request, and correlate against in a reply), preferably without making a service worker call to generate one (or a HTTP call to an online generator as described here https://forum.camunda.io/t/built-in-function-for-uuid/35018/14).

So another useful addition could be to generate this, for example by calling a `UUID()` function. An alternative to this could be to use the `camunda.processInstanceKey` when that's available and add a suffix when you need multiple ID's in your process.

**Comment by nikku** [comment]
Related to [SUPPORT-18461](https://jira.camunda.com/browse/SUPPORT-18461).

**Comment by edfred99** [comment]
I need that function too to built a much smoother integration with the existing application landscape

**Comment by igor-msk** [comment]
I'm very interested in this feature too

**Comment by capobandito** [comment]
plus one for this, we would like to use it

**Comment by ekzobrain** [comment]
+1

**Comment by cschaf** [comment]
+1

**Comment by misslisserw** [comment]
+1

**Comment by valeriaportolesicamunda** [comment]
Feature requested in [SUPPORT-25550](https://jira.camunda.com/browse/SUPPORT-25550) as well.

**Comment by korthout** [comment]
During the 2025 kickoff, a few of us hacked on this topic to explore the different use cases and the implementation challenges. You can find our results in #27527.

To push this topic along, I'm working on a breakdown of the feature so we can plan our work on this accordingly.

While we could build this incrementally, we need some upfront design to ensure that we provide a holistic vision for the feature. For example by:
- considering which contexts we want to provide eventually (based on use cases), align naming, and group concepts
- where those contexts will be available (in all expressions across the product or for specific expressions only), and whether the data is available across the involved components
- what infrastructure do we need to support this
- what we need to offer a great user experience (documentation, modeler suggestions, linting, co-pilot, etc)

I'd like to take the following steps:

1. Let's first define the entire Camunda context as we see it right now. 
2. We can then consider in which expressions these contexts and their properties can be made available, and what use cases they solve. 
3. We look at technical challenges per context/property to help us prioritize high-impact/low-effort cases first.
4. Breakdown the properties we want to support into separate issues
5. Breakdown the necessary required infrastructure changes
6. Breakdown the needed changes to offer a great UX

**The Camunda context**

What follows is the definition of the camunda expression context as we see it in scope right now.

```json
"camunda": {
  "env": {
    "CAMUNDA_MY_ENV_VAR": "Env var with camunda prefix",
    "CUSTOM_MY_ENV_VAR": "Env var with a custom prefix"
  },
  "error": {
    "code": "The code of a caught error",
    "message": "The message of a caught error"
  },
  "job": {
    "retries": "The number of retries left for this job"
  },
  "multiInstance": {
    "numberOfInstances": "The number of instances created",
    "numberOfActiveInstances": "The number of instances currently active",
    "numberOfCompletedInstances": "The number of instances completed",
    "numberOfTerminatedInstances": "The number of instances terminated"
  },
  "process": {
    "elements": [{
      "id": "The element ID",
      "type": "The element type",
      "documentation": "The element documentation",
      "adHocActivities": ["The ID of each activatable activity inside of an ad hoc sub-process"]
    }],
    "versionTag": "The version tag of the process"
  },
  "processInstance": {
    "key": "The process instance key"
  },
  "userTask": {
    "assignee": "The user task's assignee",
    "candidateGroups": "The user task's candidate groups",
    "candidateUsers": "The user task's candidate users",
    "dueDate": "The user task's due date",
    "followUpDate": "The user task's follow-up date",
    "priority": "The user task's priority"
  }
}
```

<details><summary>Not yet in scope</summary>
<p>

These are just ideas, but we have not matched use cases yet.

```json
"camunda": {
  "process": {
    "key": "The process definition key",
    "id": "The process ID",
    "version": "The process version"
  },
  "processInstance": {
    "parentProcessInstanceKey": "The process instance key of the parent process instance"
  },
  "error": {
    "variables": "The variables of a caught error"
  },
  "elementInstance": {
    "key": "The element instance key",
    "elementId": "The element ID of this element instance",
    "elementType": "The element type of this element instance"
  },
  "cluster": {
    "id": "The cluster ID",
    "name": "The cluster name",
    "tag": "The cluster tag"
  }
}
```

</p>
</details> 

**Next steps**

This was just step 1 from the steps above. I'll continue with step 2 next.

**Comment by placsta** [comment]
+1

**Comment by mschulte** [comment]
Good to see this one evolve @korthout .

We would need this too in a project for an enterprise customer. 
Currently we plan to build our own generic worker propagating camunda entity attributes (as we are missing that feature)

**Comment by renzpatriarca** [comment]
Customer also need this feature in [SUPPORT-26876](https://jira.camunda.com/browse/SUPPORT-26876)

**Comment by korthout** [comment]
After some internal feedback, We've re-considered use cases. 

We'll scope out the following properties for now:

- `camunda.job.retries` (we have a use case, but the case requires additional functionality)
- `camunda.userTask.candidateGroups` 
- `camunda.userTask.candidateUsers`
- `camunda.userTask.priority`

We also want to scope in the following properties:

- `camunda.process.elements.name`
- `camunda.process.key`
- `camunda.process.id`
- `camunda.process.version`

We also want to adjust the `camunda.process.elements.adHocActivities` property to provide a list of all properties of the element, rather than just the ID.

The new scope is then:

```json
"camunda": {
  "env": {
    "CAMUNDA_MY_ENV_VAR": "Env var with camunda prefix",
    "CUSTOM_MY_ENV_VAR": "Env var with a custom prefix"
  },
  "error": {
    "code": "The code of a caught error",
    "message": "The message of a caught error"
  },
  "multiInstance": {
    "numberOfInstances": "The number of instances created",
    "numberOfActiveInstances": "The number of instances currently active",
    "numberOfCompletedInstances": "The number of instances completed",
    "numberOfTerminatedInstances": "The number of instances terminated"
  },
  "process": {
    "key": "The process definition key",
    "id": "The process ID",
    "name": "The process name",
    "version": "The process version"
    "versionTag": "The version tag of the process",
    "elements": [{
      "id": "The element ID",
      "name": "The element name"
      "type": "The element type",
      "documentation": "The element documentation",
      "adHocActivities": [{
        "id": "The element ID of an activatable activity inside an ad-hoc sub-process",
        "name": "The element name of an activatable activity inside an ad-hoc sub-process"
        "type": "The element type of an activatable activity inside an ad-hoc sub-process",
        "documentation": "The element documentation of an activatable activity inside an ad-hoc sub-process"
      }]
    }]
  },
  "processInstance": {
    "key": "The process instance key"
  },
  "userTask": {
    "assignee": "The user task's assignee",
    "dueDate": "The user task's due date",
    "followUpDate": "The user task's follow-up date"
  }
}
```

We can split this into use cases (please let me know if you have a use case that is not covered here):

1. Access the process and process instance attributes like processInstanceKey and process name (this issue) - available in input mapping expressions
2. cluster-specific env vars (https://github.com/camunda/product-hub/issues/637) - available in input mapping expressions
3. Capture error code and message when catching bpmn error (https://github.com/camunda/camunda/issues/11459) - available in output mapping expressions on error catch event
4. Align [multi-instance completion condition properties](https://docs.camunda.io/docs/next/components/modeler/bpmn/multi-instance/#completion-condition) - available in completion condition expression on multi-instance marked element
5. Flexible [ad-hoc sub-process](https://docs.camunda.io/docs/next/components/modeler/bpmn/ad-hoc-subprocesses/) activation without need to [search for activatable activities](https://docs.camunda.io/docs/next/apis-tools/camunda-api-rest/specifications/search-ad-hoc-subprocess-activities/) - available in `activeElementsCollection` expression on ad-hoc sub-process and on input mapping expressions of any element (allows to pass the activatable activities to a job worker)
6. Output assignee of completed user task as a variable, to use in follow-up task. E.g. set next task to same assignee, or exclude assignee from next task's candidate users - available on output mapping expressions of user tasks
7. Output dueDate/followUpDate of completed user task as a variable, to use in follow-up task - available on output mapping expressions of user tasks

I'll open issues for each of these cases.

---

The rest of the ideas, for which we don't have matched use cases yet:

```json
"camunda": {
  "processInstance": {
    "parentProcessInstanceKey": "The process instance key of the parent process instance"
  },
  "error": {
    "variables": "The variables of a caught error"
  },
  "job": {
    "retries": "The number of retries left for this job"
  },
  "userTask": {
    "candidateGroups": "The user task's candidate groups",
    "candidateUsers": "The user task's candidate users",
    "priority": "The user task's priority"
  },
  "elementInstance": {
    "key": "The element instance key",
    "elementId": "The element ID of this element instance",
    "elementType": "The element type of this element instance"
  },
  "cluster": {
    "id": "The cluster ID",
    "name": "The cluster name",
    "tag": "The cluster tag"
  }
}
```

**Comment by edfred99** [comment]
Hi, a possibility to get the current userTask id so that I can pass it over a link to an other application which completes the task . This will reduce the need of some boilerplate code proven id‘s which pollute the model. It will also make the glue code at the App side less complex (e.g. filter tasklist for appropriate items)

**Comment by saig0** [comment]
@korthout let's remove the ad-hoc sub-process property (`camunda.process.elements.adHocActivities`) from the scope. Instead of the property, we will work on direct support for job-workers (and user tasks) to control the ad-hoc sub-process. So, we don't have a strong need for this property anymore.

**Comment by jonathanlukas** [comment]
Quick thought from my side: 90%+ of the usage of the context will be to retrieve the `processInstanceKey`, how about making this easily accessible:
```json
"camunda": {
  "processInstanceKey": "process instance key"
}
```

**Comment by korthout** [comment]
@jonathanlukas In usage that's just one character fewer, at the cost of consistency. Is that really that much easier?

```
camunda.processInstance.key
camunda.processInstanceKey
```

💡 If you think many users may struggle to find the key this way, then let's make it easy to discover the `camunda.processInstance.key` in the intellisense of Modeler.

**Comment by jonathanlukas** [comment]
@korthout which other attributes will be mapped to the 
camunda.processInstance?

**Comment by korthout** [comment]
I imagine other metadata. For example, tags and start date would be useful to access.

**Comment by jonathanlukas** [comment]
I agree, this makes sense 🙌

**Comment by matthewschwarze95** [comment]
Feature requested in [SUPPORT-30320](https://jira.camunda.com/browse/SUPPORT-30320) as well

**Comment by jphillips-bp3** [comment]
This would be useful from a C7 to C8 migration point of view, where correlation messages in C7 can be correlated against the PID, but in C8 it is a correlation key.

In this case, we could use the C8 PIK as the correlation key, rather than having to generate another Id to correlation against. The customer can continue to use the fields that they store the C7 PID's with C8 PIK's.

For example:

<img width="334" height="372" alt="Image" src="https://github.com/user-attachments/assets/eb759284-03c0-4b92-aec3-e9d43acd5c31" />

**Comment by ev-codes** [comment]
@korthout @mustafadagher It would also be helpful to have the process instance's business ID available.

That way, [modelers](https://github.com/camunda/camunda/issues/51694) could specify child instance business ID using the parent's business ID as a prefix (e.g., `=camunda.processInstance.businessId + "-child-suffix"`
